### PR TITLE
feat: add 'captureBody' support for Hapi

### PR DESCRIPTION
**Pull request by trentm**
The body of incoming requests to an HTTP server using the Hapi
framework will be recorded in relevant APM transaction and error
objects depending on the `captureBody` config var setting.

Closes: #1905

### Checklist

- [x] Implement code
- [x] Add tests
- [x] Add CHANGELOG.asciidoc entry
- [x] Commit message follows [commit guidelines](https://github.com/elastic/apm-agent-nodejs/blob/main/CONTRIBUTING.md#commit-message-guidelines)

### CODE REVIEW

**Comment by trentm** [comment]
> removed the request for review ...

I have a test hang with node v16 and v17 to solve first.

**Comment by trentm** [comment]
run module tests for hapi,@hapi/hapi

**Comment by trentm** [comment]
TAV run here: https://apm-ci.elastic.co/job/apm-agent-nodejs/job/apm-agent-nodejs-mbp/job/PR-2618/3/

**Comment by mshustov** [comment]
@trentm LGTM. I'm going to test it with Kibana

**Comment by trentm** [comment]
> LGTM. Tested with Kibana, it works fine. 👌

@mshustov  Good to hear. 
I still have a failure in our "TAV" (test all versions) tests with node v16 and v17 when testing with `@hapi/hapi@20.1.1` that I need to understand. `@hapi/hapi@20.1.2` and newer passed the tests.

**Comment by trentm** [comment]
The TAV run failed https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-nodejs%2Fapm-agent-nodejs-mbp/detail/PR-2618/3/pipeline
The failures were with testing `@hapi/hapi` with node 16 and 17, starting
with `@hapi/hapi@20.1.1`.

```
[2022-03-23T00:55:38.851Z] node_tests_1  | -- required packages ["@hapi/hapi@20.1.2"]
[2022-03-23T00:55:38.851Z] node_tests_1  | -- installing ["@hapi/hapi@20.1.2"]
[2022-03-23T00:55:43.059Z] node_tests_1  | -- running test "node test/instrumentation/modules/hapi/basic.test.js" with @hapi/hapi
[2022-03-23T00:55:43.319Z] node_tests_1  | -- running test "node test/instrumentation/modules/hapi/set-framework-hapihapi.test.js" with @hapi/hapi
[2022-03-23T00:55:43.889Z] node_tests_1  | -- required packages ["@hapi/hapi@20.1.1"]
[2022-03-23T00:55:43.889Z] node_tests_1  | -- installing ["@hapi/hapi@20.1.1"]
[2022-03-23T00:55:48.145Z] node_tests_1  | -- running test "node test/instrumentation/modules/hapi/basic.test.js" with @hapi/hapi
[2022-03-23T00:55:48.145Z] node_tests_1  | /app/node_modules/@hapi/hapi/lib/request.js:499
[2022-03-23T00:55:48.145Z] node_tests_1  |             if (this.response.statusCode === 500 &&
[2022-03-23T00:55:48.145Z] node_tests_1  |                               ^
[2022-03-23T00:55:48.145Z] node_tests_1  |
[2022-03-23T00:55:48.145Z] node_tests_1  | TypeError: Cannot read properties of null (reading 'statusCode')
[2022-03-23T00:55:48.145Z] node_tests_1  |     at Request._finalize (/app/node_modules/@hapi/hapi/lib/request.js:499:31)
[2022-03-23T00:55:48.145Z] node_tests_1  |     at Request._reply (/app/node_modules/@hapi/hapi/lib/request.js:436:18)
[2022-03-23T00:55:48.145Z] node_tests_1  |     at Request._execute (/app/node_modules/@hapi/hapi/lib/request.js:281:14)
[2022-03-23T00:55:48.145Z] node_tests_1  |     at processTicksAndRejections (node:internal/process/task_queues:96:5)
[2022-03-23T00:55:48.145Z] node_tests_1  | -- detected failing command, flushing stdout...
[2022-03-23T00:55:48.145Z] node_tests_1  | TAP version 13
[2022-03-23T00:55:48.145Z] node_tests_1  | # extract URL from request
[2022-03-23T00:55:48.146Z] node_tests_1  | ok 1 no error from startServer
[2022-03-23T00:55:48.146Z] node_tests_1  | ok 2 should be strictly equal
[2022-03-23T00:55:48.146Z] node_tests_1  | ok 3 should be strictly equal
[2022-03-23T00:55:48.146Z] node_tests_1  | ok 4 should be strictly equal
[2022-03-23T00:55:48.146Z] node_tests_1  | ok 5 should be strictly equal
[2022-03-23T00:55:48.146Z] node_tests_1  | ok 6 should be strictly equal
[2022-03-23T00:55:48.146Z] node_tests_1  | ok 7 should be strictly equal
[2022-03-23T00:55:48.146Z] node_tests_1  | ok 8 should be strictly equal
[2022-03-23T00:55:48.146Z] node_tests_1  | ok 9 should be strictly equal
[2022-03-23T00:55:48.146Z] node_tests_1  | # route naming
[2022-03-23T00:55:48.146Z] node_tests_1  | ok 10 null
[2022-03-23T00:55:48.146Z] node_tests_1  | ok 11 should be strictly equal
[2022-03-23T00:55:48.146Z] node_tests_1  | ok 12 should be strictly equal
[2022-03-23T00:55:48.146Z] node_tests_1  | ok 13 should be strictly equal
[2022-03-23T00:55:48.146Z] node_tests_1  | ok 14 should be strictly equal
[2022-03-23T00:55:48.146Z] node_tests_1  | ok 15 should be strictly equal
[2022-03-23T00:55:48.146Z] node_tests_1  | ok 16 should be strictly equal
[2022-03-23T00:55:48.146Z] node_tests_1  | ok 17 should be strictly equal
[2022-03-23T00:55:48.146Z] node_tests_1  | # captureBody
[2022-03-23T00:55:48.146Z] node_tests_1  | ok 18 null
[2022-03-23T00:55:48.146Z] node_tests_1  |
[2022-03-23T00:55:48.146Z] node_tests_1  | -- fatal: Test exited with code 1
```

This failure happens even if I disable the added `captureBody` processing,
and also when disabling the agent.  The issue occurs for a Hapi server handling
a POST (rather than just a GET, which is all our tests were testing before
this). `@hapi/hapi@20.1.2` fixes the issue https://github.com/hapijs/hapi/pull/4225
Any earlier happy version (at least back to 18.x), hits the same issue.

I.e. to use `@hapi/hapi` with node v16 or later requires at least `20.1.2`.

Handled in https://github.com/elastic/apm-agent-nodejs/pull/2618/commits/3a787b48e304377d5072a451fcae5fc5c1b239e6

**Comment by trentm** [comment]
TAV tests passed: https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-nodejs%2Fapm-agent-nodejs-mbp/detail/PR-2618/5/pipeline/

**Review by astorm — APPROVED** [review]
Code looks like it does the thing.  I also spun up a quick HAPI test app with the following route

```
    server.route({
      method: 'POST',
      path: '/test-post',
      handler: (request, h) => {
          console.log(request.payload)
          return 'Hello World!' + "\n" + JSON.stringify(request.payload);
      }
    });
```

and posted via `curl localhost:3000/test-post -d 'foo=bar'`, and the request body made it through to the `http.request.body.original` field in the UI

thereitis


Approving

**Comment by trentm** [comment]
> the request body made it through to the http.request.body.original field in the UI

@astorm Whoa what is going on with the `http.request.body.original` value in your screenshot. It looks like there is an extra single-quote on the "foo": `{"'foo": "bar"}`. Am I misreading that?

**Comment by astorm** [comment]
> Oh there is a trailing single-quote on the "bar" as well. So I'm guessing that curl call included the literal single-quotes for the form-encoded POST body?

Good question.  This looks like a weird side effect of shell escaping in a bash script I used for testing and a type 1 thinking copy-pasta.  I call the script like this

```
curl-sleep.bash "localhost:3000/test-post -d 'foo=bar'"
```

and `curl-sleep` uses that like `curl $1`, so those single quotes ended up included as literals.  FWIW a regular invocation of curl 

```
curl localhost:3000/test-post -d 'foo=bar'
```

does not produce those extra quotes. 

TL;DR; -- all clear :)
